third_party/terraform/utils/storage_transfer_helpers.go

```diff
@@ -0,0 +1,312 @@
+	for _, data := range options {
+		datasSchema = append(datasSchema, map[string]interface{}{
+			"delete_objects_from_source_after_transfer":  data.DeleteObjectsFromSourceAfterTransfer,
+			"delete_objects_unique_in_sink":              data.DeleteObjectsUniqueInSink,
+			"overwrite_objects_already_existing_in_sink": data.OverwriteObjectsAlreadyExistingInSink,
+		})
+	}
+	return datasSchema
+}
+
+func expandTransferSpecs(transferSpecs []interface{}) []*storagetransfer.TransferSpec {
+	specs := make([]*storagetransfer.TransferSpec, 0, len(transferSpecs))
+	for _, raw := range transferSpecs {
+		spec := raw.(map[string]interface{})
+
+		transferSpec := &storagetransfer.TransferSpec{
+			GcsDataSink: expandGcsData(spec["gcs_data_sink"].([]interface{}))[0],
+		}
+
+		if v, ok := spec["object_conditions"]; ok && len(v.([]interface{})) > 0 {
+			transferSpec.ObjectConditions = expandObjectConditions(v.([]interface{}))[0]
+		}
+		if v, ok := spec["transfer_options"]; ok && len(v.([]interface{})) > 0 {
+			transferSpec.TransferOptions = expandTransferOptions(v.([]interface{}))[0]
+		}
+
+		if v, ok := spec["gcs_data_source"]; ok && len(v.([]interface{})) > 0 {
+			transferSpec.GcsDataSource = expandGcsData(v.([]interface{}))[0]
+		} else if v, ok := spec["aws_s3_data_source"]; ok && len(v.([]interface{})) > 0 {
+			transferSpec.AwsS3DataSource = expandAwsS3Data(v.([]interface{}))[0]
+		} else if v, ok := spec["http_data_source"]; ok && len(v.([]interface{})) > 0 {
+			transferSpec.HttpDataSource = expandHttpData(v.([]interface{}))[0]
+		}
+
+		specs = append(specs, transferSpec)
+	}
+	return specs
+}
+
+func flattenTransferSpecs(transferSpecs []*storagetransfer.TransferSpec) []map[string][]map[string]interface{} {
+	transferSpecsSchema := make([]map[string][]map[string]interface{}, 0, len(transferSpecs))
+	for _, transferSpec := range transferSpecs {
+		schema := map[string][]map[string]interface{}{
+			"gcs_data_sink": flattenGcsData([]*storagetransfer.GcsData{transferSpec.GcsDataSink}),
+		}
+
+		if transferSpec.ObjectConditions != nil {
+			schema["object_conditions"] = flattenObjectConditions([]*storagetransfer.ObjectConditions{transferSpec.ObjectConditions})
+		}
+		if transferSpec.TransferOptions != nil {
+			schema["transfer_options"] = flattenTransferOptions([]*storagetransfer.TransferOptions{transferSpec.TransferOptions})
+		}
+		if transferSpec.GcsDataSource != nil {
+			schema["gcs_data_source"] = flattenGcsData([]*storagetransfer.GcsData{transferSpec.GcsDataSource})
+		} else if transferSpec.AwsS3DataSource != nil {
+			schema["aws_s3_data_source"] = flattenAwsS3Data([]*storagetransfer.AwsS3Data{transferSpec.AwsS3DataSource})
+		} else if transferSpec.HttpDataSource != nil {
+			schema["http_data_source"] = flattenHttpData([]*storagetransfer.HttpData{transferSpec.HttpDataSource})
+		}
+
+		transferSpecsSchema = append(transferSpecsSchema, schema)
+	}
+	return transferSpecsSchema
+}
+
+func validateDuration() schema.SchemaValidateFunc {
```

Comment:
nit: this can live in [`validation.go`](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/utils/validation.go)